### PR TITLE
CIRC-660: Resolve claimed returned item status via check-in

### PR DESCRIPTION
**Pull request by bohdan-suprun**
https://issues.folio.org/browse/CIRC-660 - Claimed returned: check in item

## Approach
- Add new `claimedReturnedResolvedBy` enum property for check-in-by-barcode API;
- Add validators;
- Introduce new loan actions:
   - `checkedInFoundByLibrary`;
   - `checkedInReturnedByPatron`

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/LoanAction.java [hunk 1]
```diff
@@ -20,4 +20,19 @@
   public String getValue() {
     return value;
   }
+
+  public enum ResolveClaimedReturned {
+    CHECKED_IN_RETURNED_BY_PATRON("checkedInReturnedByPatron"),
+    CHECKED_IN_FOUND_BY_LIBRARY("checkedInFoundByLibrary");
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/LoanAction.java]: What is the significance of `checkedIn` as a prefix for these values? it doesn't seem to add much information over the rest of the text.

File: src/main/java/org/folio/circulation/domain/LoanCheckInService.java [hunk 2]
```diff
@@ -1,28 +1,53 @@
...
+          request.getCheckInDate(), systemDateTime, request.getServicePointId()));
+    }
+
+    return succeeded(loan.checkIn(request.getCheckInDate(), systemDateTime,
+      request.getServicePointId()));
   }
 
   public boolean isInHouseUse(Item item, RequestQueue requestQueue,
-    CheckInByBarcodeRequest checkInByBarcodeRequest) {
+                              CheckInByBarcodeRequest checkInByBarcodeRequest) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/LoanCheckInService.java]: Please undo this change to the indentation

File: src/main/java/org/folio/circulation/domain/LoanCheckInService.java [hunk 3]
```diff
@@ -1,28 +1,53 @@
...
+    switch (request.getClaimedReturnedResolvedBy()) {
+      case FOUND_BY_LIBRARY:
+        return ResolveClaimedReturned.CHECKED_IN_FOUND_BY_LIBRARY;
+      case RETURNED_BY_PATRON:
+        return ResolveClaimedReturned.CHECKED_IN_RETURNED_BY_PATRON;
+      default:
+        throw new IllegalArgumentException(
+          "No loan action to resolve claimed returned: "
+            + request.getClaimedReturnedResolvedBy());
+    }
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/LoanCheckInService.java]: What catches this exception?

File: src/main/java/org/folio/circulation/domain/validation/CheckInValidators.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  public static Result<CheckInProcessRecords> refuseWhenClaimedReturnedIsNotResolved(
+    Result<CheckInProcessRecords> processRecordsResult) {
+
+    return processRecordsResult.failWhen(
+      processRecords -> succeeded(isClaimedReturnedNotResolved(processRecords)),
+      processRecords -> singleValidationError(String.format(
+        "Item has Claimed returned status but no '%s' present", CLAIMED_RETURNED_RESOLVED_BY),
+        CLAIMED_RETURNED_RESOLVED_BY, null));
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/validation/CheckInValidators.java]: Why is this message parameterised? If this is presented to the user, will they understand what the name of the JSON property mean?

File: src/test/java/api/support/builders/CheckInByBarcodeRequestBuilder.java [hunk 5]
```diff
@@ -34,6 +39,7 @@ public JsonObject create() {
     put(request, "itemBarcode", this.itemBarcode);
     put(request, "checkInDate", this.checkInDate);
     put(request, "servicePointId", this.servicePointId);
+    put(request, CLAIMED_RETURNED_RESOLVED_BY, this.claimedReturnedResolvedBy);
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/builders/CheckInByBarcodeRequestBuilder.java]: Is this constant from the production code?

File: ramls/check-in-by-barcode-request.json [hunk 6]
```diff
@@ -16,6 +16,14 @@
...
       "type": "string",
       "format": "date-time"
+    },
+    "claimedReturnedResolvedBy": {
+      "description": "Defines who has resolved claimed returned item state",
+      "type": "string",
+      "enum": [
+        "Found by library",
+        "Returned by patron"
+      ]
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply ramls/check-in-by-barcode-request.json]: I think we should proceed with the kind of resolution, should I rename it to: `claimedReturnedResolutionReason`? Any other suggestions?

File: src/main/java/org/folio/circulation/domain/LoanAction.java (hunk 1 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: This action was mentioned in https://issues.folio.org/browse/UIU-1218:
image

I've decided to use similar actions names.

File: src/main/java/org/folio/circulation/domain/LoanCheckInService.java (hunk 3 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/LoanCheckInService.java]: It should be interpreted to a server error, I think. The main purpose of the exception is to notify early if there is a new claimed returned resolution type but it has not mapped to a specific loan action.

File: src/main/java/org/folio/circulation/domain/validation/CheckInValidators.java (hunk 4 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/validation/CheckInValidators.java]: The property name should be self-explanatory enough. Do you have an idea how it can be reworded?

File: src/test/java/api/support/builders/CheckInByBarcodeRequestBuilder.java (hunk 5 above)
Comment (bohdan-suprun) [reply src/test/java/api/support/builders/CheckInByBarcodeRequestBuilder.java]: Yes, it is from the CheckInByBarcodeRequest class.

File: ramls/check-in-by-barcode-request.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcjohnson-kint) [reply ramls/check-in-by-barcode-request.json]: > I think we should proceed with the kind of resolution, should I rename it to: `claimedReturnedResolutionReason`? Any other suggestions?

My suggestion would be `claimedReturnedResolution`. 

In conversations with Emma, how does she refer to this, does she call it a resolution or a reason or something else? It might be worth asking her.

File: src/main/java/org/folio/circulation/domain/LoanAction.java (hunk 1 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: Ok. The way actions are modelled is a bit of debt related to how the action history is created, it is a bit messy.

Why did you use that formatting for the actions rather than the text that Emma asked for directly? Is it because the UI would prefer to use a code / constant like value that can then be translated?

File: src/main/java/org/folio/circulation/domain/LoanCheckInService.java (hunk 3 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanCheckInService.java]: Please can the calling code for this method be checked that it does interpret this exception as a server error.

File: src/main/java/org/folio/circulation/domain/validation/CheckInValidators.java (hunk 4 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/validation/CheckInValidators.java]: Personally, I prefer the messages to be in domain language where possible. I might suggest something like `Item is claimed returned, a resolution is required to check in` or something like that, depending upon the language used by Emma or other domain experts.

File: src/test/java/api/support/builders/CheckInByBarcodeRequestBuilder.java (hunk 5 above)
Comment (marcjohnson-kint) [reply src/test/java/api/support/builders/CheckInByBarcodeRequestBuilder.java]: I would prefer that the API test code does not depend upon production code. I think this kind of reuse weakens the tests, partly for readability / documentation and partly because the tests won't  identify when the API has changed or the constant has become incorrect.

File: src/main/java/org/folio/circulation/domain/LoanAction.java (hunk 1 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: I just followed existing approach for loan actions. But I agree, that setting the text is clearer than having a constant and then have a mapping on UI. Should we start using text for these loan actions?

File: ramls/check-in-by-barcode-request.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply ramls/check-in-by-barcode-request.json]: Renamed to `claimedReturnedResolution`

File: src/main/java/org/folio/circulation/domain/validation/CheckInValidators.java (hunk 4 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/validation/CheckInValidators.java]: Changed to `Item is claimed returned, a resolution is required to check in`

File: src/main/java/org/folio/circulation/domain/LoanCheckInService.java (hunk 3 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/LoanCheckInService.java]: Unfortunately it is not possible to test this, because we have to add a new value for `ClaimedReturnedResolution` enum. which is not possible to achieve through tests. 

I've changed it to return a result instead.

File: ramls/check-in-by-barcode-request.json [hunk 7]
```diff
@@ -16,6 +16,14 @@
       "description": "When the item was checked in",
       "type": "string",
       "format": "date-time"
+    },
+    "claimedReturnedResolution": {
+      "description": "Defines why the claimed returned state of item has been resolved",
```
Comment (marcjohnson-kint) [inline ramls/check-in-by-barcode-request.json]:
```suggestion
      "description": "Describes how the library resolved the situation where an item that was claimed returned",
```

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 8]
```diff
@@ -744,6 +752,120 @@ private void checkPatronNoticeEvent(
...
+        .at(checkInServicePointId));
+
+    assertThat(checkInResponse.getJson(), hasErrorWith(allOf(
+      hasMessage("Item is claimed returned, a resolution is required to check in"),
+      hasNullParameter("claimedReturnedResolvedBy")
+    )));
+  }
+
+  @Test
+  public void cannotResolveClaimedReturnedIfWrongResolvedByValueProvided() {
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/CheckInByBarcodeTests.java]:
```suggestion
  public void cannotResolveClaimedReturnedIfUnrecognisedResolutionProvided() {
```

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 9]
```diff
@@ -744,6 +752,120 @@ private void checkPatronNoticeEvent(
...
+        .claimedReturnedResolution(resolutionValue));
+
+    assertThat(checkInResponse.getJson(), hasErrorWith(allOf(
+      hasMessage("Unrecognized value provided for property"),
+      hasParameter("claimedReturnedResolution", resolutionValue)
+    )));
+  }
+
+  @Test
+  public void claimedReturnedResolvedByPropertyIsIgnoredForCheckedOutItem() {
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/CheckInByBarcodeTests.java]:
```suggestion
  public void claimedReturnedResolvedByPropertyIsIgnoredForItemNotClaimedReturned() {
```

File: src/main/java/org/folio/circulation/domain/LoanAction.java (hunk 1 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: I think we should ask the UI folks which they prefer. Both styles are used at the moment I believe.

File: src/main/java/org/folio/circulation/domain/LoanCheckInService.java (hunk 3 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanCheckInService.java]: I didn't necessarily mean an automated test, rather an inspection of the code paths where this was used.

File: src/main/java/org/folio/circulation/domain/LoanAction.java (hunk 1 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: Hi @marcjohnson-kint, I've checked it with @SergiyVSergiyenko, both approaches will require mapping, because such values have to be translate for internationalization, it is just about what key to use for a value.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: Hmm, ok. Out of interest, how do we translate dynamic values like this?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: If they are equivalent, I suggest we leave this as is.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: (Actions and loan action history suffer from significant debt due to how they were originally built, this is a bit of a crutch)

Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/domain/LoanAction.java]: There are translations files, with key-value pairs: https://github.com/folio-org/ui-users/blob/master/translations/ui-users/en.json#L29-L39
